src/scripts/OSFramework/OSUI/Feature/Balloon/Balloon.ts

```diff
@@ -0,0 +1,332 @@
+		public isOpen = false;
+
+		constructor(featurePattern: PT, featureElem: HTMLElement, options: BalloonOptions) {
+			super(featurePattern, featureElem, options);
+			this.build();
+		}
+
+		// Method to handle the body click callback, that closes the Balloon
+		private _bodyClickCallback(_args: string, e: MouseEvent): void {
+			if (e.target === this.featureOptions?.anchorElem || this._isOpenedByApi) {
+				return;
+			}
+			if (this.isOpen) {
+				this._toggleBalloon(false, true);
+				e.stopPropagation();
+			}
+		}
+
+		// Add Focus Trap to the Pattern
+		private _handleFocusTrap(): void {
+			const opts = {
+				focusTargetElement: this._floatingUIOptions.anchorElem.parentElement,
+			} as Behaviors.FocusTrapParams;
+
+			this._focusTrapInstance = new Behaviors.FocusTrap(opts);
+		}
+
+		// Call methods to open or close, based on e.key and behaviour applied
+		private _onkeypressCallback(e: KeyboardEvent): void {
+			const isEscapedPressed = e.key === GlobalEnum.Keycodes.Escape;
+
+			// Close the Balloon when pressing Esc
+			if (isEscapedPressed && this.isOpen) {
+				this.close();
+			}
+		}
+
+		// Method to remove the event listeners
+		private _removeEventListeners(): void {
+			this.featureElem.removeEventListener(GlobalEnum.HTMLEvent.keyDown, this._eventOnKeypress);
+			Event.DOMEvents.Listeners.GlobalListenerManager.Instance.removeHandler(
+				Event.DOMEvents.Listeners.Type.BodyOnClick,
+				this._eventBodyClick
+			);
+		}
+
+		// Add the Accessibility Attributes values
+		// eslint-disable-next-line @typescript-eslint/naming-convention
+		private _setA11YProperties(): void {
+			Helper.Dom.Attribute.Set(this.featureElem, Constants.A11YAttributes.Aria.Hidden, (!this.isOpen).toString());
+
+			// Will handle the tabindex value of the elements inside pattern
+			Helper.A11Y.SetElementsTabIndex(this.isOpen, this._focusTrapInstance.focusableElements);
+
+			Helper.Dom.Attribute.Set(
+				this.featureElem,
+				Constants.A11YAttributes.TabIndex,
+				this.isOpen
+					? Constants.A11YAttributes.States.TabIndexShow
+					: Constants.A11YAttributes.States.TabIndexHidden
+			);
+
+			Helper.Dom.Attribute.Set(
+				this._floatingUIOptions.anchorElem,
+				Constants.A11YAttributes.TabIndex,
+				this.isOpen
+					? Constants.A11YAttributes.States.TabIndexHidden
+					: Constants.A11YAttributes.States.TabIndexShow
+			);
+		}
+
+		// Set the callbacks
+		private _setCallbacks(): void {
+			this._eventBodyClick = this._bodyClickCallback.bind(this);
+			this._eventOnKeypress = this._onkeypressCallback.bind(this);
+
+			// Set custom Balloon event
+			this._onToggleEvent = function dispatchCustomEvent(isOpen, balloonElem) {
+				const _customEvent = new CustomEvent(GlobalEnum.CustomEvent.BalloonOnToggle, {
+					detail: { isOpen: isOpen, balloonElem: balloonElem },
+				});
+				document.dispatchEvent(_customEvent);
+			};
+
+			// Set its reference on the window
+			window[OSFramework.OSUI.GlobalEnum.CustomEvent.BalloonOnToggle] =
```

Comment:
Nothing, this is just to create the reference on the window for the custom event name, so that later we can validate it on the addEvent method from the AbstractListener.

That being said, I added an if validation, to only set if is undefined.

Reply:
Changed this to the AbstractListener